### DISCUSSION TITLE
Is it treated different when passing a string literal to the external array?

### DISCUSSION THREAD

**Discussion by danielo515** [tree-sitter/tree-sitter#1893] [Q&A] [answered]
So, several grammars (mine included) have an external scanner and declare certain nodes to be parsed by that scanner. However, some grammars also pass literal strings, that are also used latter in the grammar itself to define tokens and other nodes.
Is this any different from the regular passing of a node like `$.comment` and so?
If it is different, what are the differences?

**Comment by ahelwer** [comment, marked as answer]
It's a shorthand for
```js
externals: $ => [$.some_token]

rules: {
  some_token_rule: $ => alias($.some_token, 'some token')
}
```
so you can just write
```js
externals: $ => ['some token']

rules: {
  some_token_rule: $ => 'some token'
}
```
I have a pending PR here documenting this feature here: https://github.com/tree-sitter/tree-sitter/pull/1612

**Reply by danielo515** [reply]
Thanks for your answer. Reading other discussions about this I guet the impression that, if you pass it as a literal string, and the external scanner does not recognize it as a token it will try some fallback parsing. Is that the case?
By the way, how is that you can reference nodes that you never created? Does tree sitter do some meta programming inspection to extract those or how?

**Reply by ahelwer** [reply]
If the external scanner returns false then tree-sitter will try to match whatever tokens are valid in that location. If none match then it will enter error recovery mode. I don't think it will ever look for the literal terminal `'some token'`, which can of course be literally any string that your external scanner decides to tokenize as that (however that behavior would be confusing and usually I see this feature just used to keep the external scanner state in sync with the regular scanner state by sending it terminal keywords). I am not sure how exactly tree-sitter parser generation works but I think this particular thing could be done without metaprogramming, reflection, or parsing.
